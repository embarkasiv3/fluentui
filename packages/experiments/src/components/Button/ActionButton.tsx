import * as React from 'react';
import { Button } from './Button';
import { IButtonComponent, IButtonTokenReturnType } from './Button.types';
import { ButtonVariantsType } from './ButtonVariants.types';
import { FontWeights } from '../../Styling';

const baseTokens: IButtonComponent['tokens'] = (props, theme): IButtonTokenReturnType => {
  const { palette } = theme;

  return {
    backgroundColor: 'transparent',
    backgroundColorHovered: 'transparent',
    backgroundColorPressed: 'transparent',
    borderColor: 'transparent',
    color: palette.neutralPrimary,
    colorHovered: palette.themePrimary,
    colorPressed: palette.black,
    contentPadding: '0px 8px',
    height: '40px',
    iconColor: palette.neutralPrimary,
    iconColorHovered: palette.themePrimary,
    iconColorPressed: palette.black,
    textWeight: FontWeights.regular
  };
};

const disabledTokens: IButtonComponent['tokens'] = (props, theme): IButtonTokenReturnType => {
  const { palette } = theme;

  return {
    color: palette.neutralTertiary,
    colorHovered: palette.neutralTertiary,
    colorPressed: palette.neutralTertiary,
    iconColor: palette.neutralTertiary,
    iconColorHovered: palette.neutralTertiary,
    iconColorPressed: palette.neutralTertiary
  };
};

export const ActionButtonTokens: IButtonComponent['tokens'] = (props, theme): IButtonTokenReturnType => [
  baseTokens,
  props.disabled && disabledTokens
];

export const ActionButton: ButtonVariantsType = props => {
  const { text, iconProps, ...rest } = props;

  return <Button stack={{ horizontalAlign: 'start' }} content={text} icon={iconProps} tokens={ActionButtonTokens} {...rest} />;
};
