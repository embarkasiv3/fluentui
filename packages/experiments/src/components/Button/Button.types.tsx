import { IComponent, IComponentStyles, IHTMLElementSlot, ISlotProp, ISlottableProps, IStyleableComponentProps } from '../../Foundation';
import { IFontWeight, IStackSlot, ITextSlot } from 'office-ui-fabric-react';
import { IIconSlot } from '../../utilities/factoryComponents.types';
import { IBaseProps } from '../../Utilities';
import { IRawStyleBase } from '@uifabric/merge-styles/lib/IRawStyleBase';

export type IButtonComponent = IComponent<IButtonProps, IButtonTokens, IButtonStyles, IButtonViewProps>;

// These types are redundant with IButtonComponent but are needed until TS function return widening issue is resolved:
// https://github.com/Microsoft/TypeScript/issues/241
// For now, these helper types can be used to provide return type safety when specifying tokens and styles functions.
export type IButtonTokenReturnType = ReturnType<Extract<IButtonComponent['tokens'], Function>>;
export type IButtonStylesReturnType = ReturnType<Extract<IButtonComponent['styles'], Function>>;

export type IButtonSlot = ISlotProp<IButtonProps>;

export type IButtonRootElements = 'a' | 'button' | 'div';

export interface IButtonSlots {
  /**
   * Defines the root slot of the component.
   */
  root?: IHTMLElementSlot<IButtonRootElements>;

  /**
   * Defines the horizontal stack used for specifying the inner layout of the Button.
   */
  stack?: IStackSlot;

  /**
   * Defines the text that is displayed inside the Button.
   */
  content?: ITextSlot;

  /**
   * Defines the icon that is displayed next to the text inside the Button.
   */
  icon?: IIconSlot;
}

export interface IButton {
  /**
   * Sets focus to the Button.
   */
  focus: () => void;
}

export interface IButtonProps
  extends ISlottableProps<IButtonSlots>,
    IStyleableComponentProps<IButtonProps, IButtonTokens, IButtonStyles>,
    IBaseProps<IButton> {
  /**
   * Defines an href reference that, if provided, will make this component render as an anchor.
   */
  href?: string;

  /**
   * Defines whether the visual representation of the Button should be emphasized.
   * @defaultvalue false
   */
  primary?: boolean;

  /**
   * Defines whether the Button should be circular.
   * In general, circular Buttons should not specify the menu and container slots.
   * @defaultvalue false
   */
  circular?: boolean;

  /**
   * Defines whether the Button is disabled.
   * @defaultvalue false
   */
  disabled?: boolean;

  /**
   * Defines an event callback that is triggered when the Button is clicked.
   */
  onClick?: (ev: React.MouseEvent<HTMLElement>) => void;

  /**
   * Defines the aria label that the screen readers use when focus goes on the Button.
   */
  ariaLabel?: string;
}

export interface IButtonViewProps extends IButtonProps {
  /**
   * Defines a reference to the inner button.
   */
  buttonRef?: React.RefObject<HTMLButtonElement>;
}

export interface IButtonTokens {
  /**
   * Defines how far should the background extend within the Button.
   */
  backgroundClip?: IRawStyleBase['backgroundClip'];

  /**
   * Defines the background color of the Button.
   */
  backgroundColor?: string;

  /**
   * Defines the background color of the Button when it is in a hovered state.
   */
  backgroundColorHovered?: string;

  /**
   * Defines the background color of the Button when it is in an active state.
   */
  backgroundColorPressed?: string;

  /**
   * Defines the border color of the Button.
   */
  borderColor?: string;

  /**
   * Defines the border color of the Button when the focus is on it.
   */
  borderColorFocused?: string;

  /**
   * Defines the border color of the Button when it is in a hovered state.
   */
  borderColorHovered?: string;

  /**
   * Defines the border color of the Button when it is in a hovered state.
   */
  borderColorPressed?: string;

  /**
   * Defines the border radius of the Button.
   */
  borderRadius?: number | string;

  /**
   * Defines the border width of the Button.
   */
  borderWidth?: number | string;

  /**
   * Defines the text color of elements inside the Button.
   */
  color?: string;

  /**
   * Defines the text color of elements inside the Button when it is in a hovered state.
   */
  colorHovered?: string;

  /**
   * Defines the text color of elements inside the Button when it is in an active state.
   */
  colorPressed?: string;

  /**
   * Defines the padding of the Button, between the Button border and the Button contents.
   */
  contentPadding?: number | string;

  /**
   * Defines the padding of the Button, between the Button border and the Button contents, when the focus is on the Button.
   */
  contentPaddingFocused?: number | string;

  /**
   * Defines the mouse cursor to be displayed when pointing over the Button.
   */
  cursor?: string | undefined;

  /**
   * Defines a fixed height for the Button.
   */
  height?: number | string;

  /**
   * Defines the background color of the Button when in high contrast mode.
   */
  highContrastBackgroundColor?: string;

  /**
   * Defines the background color of the Button when it is in a hovered state and in high contrast mode.
   */
  highContrastBackgroundColorHovered?: string;

  /**
   * Defines the background color of the Button when it is in an active state and in high contrast mode.
   */
  highContrastBackgroundColorPressed?: string;

  /**
   * Defines the border color of the Button when in high contrast mode.
   */
  highContrastBorderColor?: string;

  /**
   * Defines the border color of the Button when it is in a hovered state and in high contrast mode.
   */
  highContrastBorderColorHovered?: string;

  /**
   * Defines the border color of the Button when it is in an active state and in high contrast mode.
   */
  highContrastBorderColorPressed?: string;

  /**
   * Defines the text color of elements inside the Button when in high contrast mode.
   */
  highContrastColor?: string;

  /**
   * Defines the text color of elements inside the Button when it is in a hovered state and in high contrast mode.
   */
  highContrastColorHovered?: string;

  /**
   * Defines the text color of elements inside the Button when it is in an active state and in high contrast mode.
   */
  highContrastColorPressed?: string;

  /**
   * Defines the icon color of the Button when in high contrast mode.
   */
  highContrastIconColor?: string;

  /**
   * Defines the icon color of the Button when it is in a hovered state and in high contrast mode.
   */
  highContrastIconColorHovered?: string;

  /**
   * Defines the icon color of the Button when it is in an active state and in high contrast mode.
   */
  highContrastIconColorPressed?: string;

  /**
   * Defines the icon color of the Button.
   */
  iconColor?: string;

  /**
   * Defines the icon color of the Button when it is in a hovered state.
   */
  iconColorHovered?: string;

  /**
   * Defines the icon color of the Button when it is in an active state.
   */
  iconColorPressed?: string;

  /**
   * Defines the size of the icon inside the Button.
   */
  iconSize?: number | string;

  /**
   * Defines the font weight of the icon inside the Button.
   */
  iconWeight?: number;

  /**
   * Defines the line height of elements inside the Button.
   */
  lineHeight?: number | string;

  /**
   * Defines a minimum height for the Button.
   */
  minHeight?: number | string;

  /**
   * Defines a minimum width for the Button.
   */
  minWidth?: number | string;

  /**
   * Defines whether or not to override any styling properties that would have been set in high contrast mode.
   */
  msHighContrastAdjust?: string;

  /**
   * Defines the outline color of the Button.
   */
  outlineColor?: string;

  /**
   * Defines the font to be used for the text inside the Button.
   */
  textFamily?: string;

  /**
   * Defines the size of the text inside the Button.
   */
  textSize?: number | string;

  /**
   * Defines the font weight of the text inside the Button.
   */
  textWeight?: IFontWeight;

  /**
   * Defines a fixed width for the Button.
   */
  width?: number | string;
}

export type IButtonStyles = IComponentStyles<IButtonSlots>;
