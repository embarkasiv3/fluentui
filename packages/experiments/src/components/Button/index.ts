export * from './Button';
export * from './Button.types';
export * from './ButtonVariants.types';
export * from './ActionButton';
export * from './DefaultButton';
export * from './PrimaryButton';
