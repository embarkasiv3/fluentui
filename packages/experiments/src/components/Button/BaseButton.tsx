import * as React from 'react';
import { Button } from './Button';
import { IButtonComponent, IButtonStylesReturnType, IButtonTokenReturnType } from './Button.types';
import { ButtonVariantsType } from './ButtonVariants.types';

const baseTokens: IButtonComponent['tokens'] = {
  backgroundColor: undefined,
  backgroundColorHovered: undefined,
  backgroundColorPressed: undefined,
  borderColor: undefined,
  borderColorHovered: undefined,
  borderColorPressed: undefined,
  borderRadius: undefined,
  borderWidth: 2,
  color: undefined,
  colorHovered: undefined,
  colorPressed: undefined,
  contentPadding: 0,
  cursor: 'default',
  minHeight: undefined,
  minWidth: undefined,
  textFamily: 'Arial',
  textSize: 13.3333,
  textWeight: 400
};

const disabledTokens: IButtonComponent['tokens'] = {
  color: 'grayText',
  colorHovered: 'grayText',
  colorPressed: 'grayText'
};

const BaseButtonTokens: IButtonComponent['tokens'] = (props, theme): IButtonTokenReturnType => [
  baseTokens,
  props.disabled && disabledTokens
];

const BaseButtonStyles: IButtonComponent['styles'] = (props, theme, tokens): IButtonStylesReturnType => {
  const { disabled } = props;

  return {
    root: {
      borderStyle: 'outset',
      overflow: 'visible',
      padding: '1px 6px',

      selectors: {
        ':active': {
          borderStyle: disabled ? 'outset' : 'inset',
          outline: 'none'
        }
      }
    }
  };
};

export const BaseButton: ButtonVariantsType = props => {
  const { text, iconProps, ...rest } = props;

  return <Button content={text} icon={{ props: iconProps }} tokens={BaseButtonTokens} styles={BaseButtonStyles} {...rest} />;
};
