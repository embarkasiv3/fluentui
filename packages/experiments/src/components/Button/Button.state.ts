import * as React from 'react';
import { IButtonProps, IButtonViewProps } from './Button.types';
import { BaseState } from '../../utilities/BaseState';

export type IButtonState = Pick<
  IButtonViewProps,
  'expanded' | 'onClick' | 'onKeyDown' | 'onMenuDismiss' | 'menuTarget' | 'onSecondaryActionClick'
>;

export class ButtonState extends BaseState<IButtonProps, IButtonViewProps, IButtonState> {
  constructor(props: ButtonState['props']) {
    super(props, {
      controlledProps: ['expanded']
    });

    this.state = {
      expanded: !!props.defaultExpanded,
      onClick: this._onClick,
      onKeyDown: this._onKeyDown,
      onSecondaryActionClick: this._onSecondaryActionClick,
      onMenuDismiss: this._onMenuDismiss,
      menuTarget: undefined
    };
  }

  private _onMenuDismiss = () => {
    this.setState({
      expanded: false
    });
  };

  private _onClick = (ev: React.MouseEvent<HTMLElement>) => {
    const { disabled, menu, split, primaryActionDisabled, onClick } = this.props;

    if (!disabled && !primaryActionDisabled) {
      if (onClick) {
        onClick(ev);

        if (ev.defaultPrevented) {
          return;
        }
      }

      if (!split && menu) {
        this.setState({
          expanded: !this.state.expanded,
          menuTarget: ev.currentTarget
        });
      }
    }
  };

  private _onKeyDown = (ev: React.KeyboardEvent<HTMLElement>) => {
    const { disabled, menu } = this.props;
    if (!disabled && menu && (ev.altKey || ev.metaKey) && ev.keyCode === 40) {
      this.setState({
        expanded: !this.state.expanded,
        menuTarget: ev.currentTarget
      });
    }
  };

  private _onSecondaryActionClick = (ev: React.MouseEvent<HTMLElement>) => {
    const { disabled, menu, split } = this.props;

    if (!disabled && menu && split) {
      this.setState({
        expanded: !this.state.expanded,
        menuTarget: ev.currentTarget
      });
    }
  };
}
