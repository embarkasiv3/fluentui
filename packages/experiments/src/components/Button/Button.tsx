import * as React from 'react';
import { createComponent } from '../../Foundation';
import { ButtonState as state } from './Button.state';
import { ButtonStyles as styles, ButtonTokens as tokens } from './Button.styles';
import { IButtonProps } from './Button.types';
import { ButtonView as view } from './Button.view';

export const Button: React.StatelessComponent<IButtonProps> = createComponent({
  displayName: 'Button',
  state,
  styles,
  tokens,
  view
});

export default Button;
