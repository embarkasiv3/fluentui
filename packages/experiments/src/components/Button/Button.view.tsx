/** @jsx withSlots */
import { ContextualMenu, Stack, DirectionalHint } from 'office-ui-fabric-react';
import { withSlots, getSlots, ISlots, IPropsWithChildren } from '../../Foundation';
import { Text } from '../../Text';
import { getNativeProps, buttonProperties } from '../../Utilities';
import { Icon } from '../../utilities/factoryComponents';

import { IButtonComponent, IButtonProps, IButtonSlots, IButtonViewProps } from './Button.types';

export const ButtonView: IButtonComponent['view'] = props => {
  const { disabled, onClick, split, ...rest } = props;

  // TODO: 'href' is anchor property... consider getNativeProps by root type
  const buttonProps = { ...getNativeProps(rest, buttonProperties) };

  const Slots = getSlots<IButtonProps, IButtonSlots>(props, {
    root: _deriveRootType(props),
    stack: Stack,
    icon: Icon,
    content: Text,
    menu: ContextualMenu,
    menuIcon: Icon,

    primaryActionContainer: Stack,
    secondaryActionContainer: 'span',
    splitDivider: 'span'
  });

  return (
    <Slots.root
      type="button" // stack doesn't take in native button props
      role="button"
      onClick={split ? undefined : onClick}
      {...buttonProps}
      aria-disabled={disabled}
    >
      {split ? renderSplitButton(props, Slots) : renderButton(props, Slots)}
    </Slots.root>
  );
};

function renderButton(props: IPropsWithChildren<IButtonViewProps>, Slots: ISlots<Required<IButtonSlots>>): JSX.Element {
  const { icon, content, children, menu: Menu, expanded, menuTarget, onMenuDismiss } = props;

  return (
    <Slots.stack horizontal as="span" gap={8} verticalAlign="center" horizontalAlign="center" verticalFill>
      {icon && <Slots.icon />}
      {content && <Slots.content />}
      {children}
      {Menu && (
        <Stack.Item>
          <Slots.menuIcon iconName="ChevronDown" />
        </Stack.Item>
      )}
      {expanded && Menu && <Slots.menu target={menuTarget} onDismiss={onMenuDismiss} items={[]} />}
    </Slots.stack>
  );
}

function renderSplitButton(props: IPropsWithChildren<IButtonViewProps>, Slots: ISlots<Required<IButtonSlots>>): JSX.Element {
  const { icon, content, children, menu: Menu, expanded, onClick, onSecondaryActionClick, menuTarget, onMenuDismiss } = props;

  return (
    <Slots.stack horizontal as="span" gap={8} verticalAlign="stretch" horizontalAlign="center" verticalFill>
      <Slots.primaryActionContainer
        horizontal
        as="span"
        gap={8}
        verticalAlign="center"
        horizontalAlign="center"
        verticalFill
        onClick={onClick}
      >
        {icon && <Slots.icon />}
        {content && <Slots.content />}
        {children}
      </Slots.primaryActionContainer>

      <Slots.splitDivider />

      <Slots.secondaryActionContainer onClick={onSecondaryActionClick}>
        <Slots.menuIcon iconName="ChevronDown" />
        {expanded && Menu && (
          <Slots.menu target={menuTarget} onDismiss={onMenuDismiss} items={[]} directionalHint={DirectionalHint.bottomRightEdge} />
        )}
      </Slots.secondaryActionContainer>
    </Slots.stack>
  );
}

function _deriveRootType(props: IButtonViewProps): keyof JSX.IntrinsicElements {
  return !!props.href ? 'a' : 'button';
}
