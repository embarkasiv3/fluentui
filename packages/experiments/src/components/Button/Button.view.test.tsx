import * as React from 'react';
import * as renderer from 'react-test-renderer';

import { Button } from './Button';
import { IButtonProps } from './Button.types';
import { CommandBar, Icon, Stack, Text } from 'office-ui-fabric-react';

const menuItems = [{ key: 'a', name: 'Item a' }, { key: 'b', name: 'Item b' }];
const buttonMenu: IButtonProps['menu'] = render => render((MenuType, props) => <MenuType {...props} items={menuItems} />);

describe('Button view', () => {
  it('renders a default Button with content correctly', () => {
    const component = renderer.create(<Button content="Default button" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a disabled default Button with content correctly', () => {
    const component = renderer.create(<Button disabled content="Disabled default button" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a primary Button with content correctly', () => {
    const component = renderer.create(<Button primary content="Primary button" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a disabled primary Button with content correctly', () => {
    const component = renderer.create(<Button primary disabled content="Disabled primary button" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders a default Button with content and an icon correctly', () => {
    const component = renderer.create(<Button icon="Upload" content="Button with string icon" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a primary Button with custom content and icon right-aligned correctly', () => {
    const component = renderer.create(
      <Button primary>
        <Text>With custom text/icon right aligned</Text>
        <Icon iconName="Upload" />
      </Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a default circular Button with an icon correctly', () => {
    const component = renderer.create(<Button icon="PeopleAdd" circular />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a disabled default circular Button with an icon correctly', () => {
    const component = renderer.create(<Button icon="Phone" circular disabled />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a primary circular Button with an icon correctly', () => {
    const component = renderer.create(<Button icon="FontSize" circular primary />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a disabled primary circular Button with an icon correctly', () => {
    const component = renderer.create(<Button icon="Attach" circular primary disabled />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a menu Button correctly', () => {
    const component = renderer.create(<Button content="Menu button" menu={buttonMenu} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a menu primary Button correctly', () => {
    const component = renderer.create(<Button primary content="Menu primary button" menu={buttonMenu} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a menu disabled Button correctly', () => {
    const component = renderer.create(<Button disabled content="Menu disabled button" menu={buttonMenu} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a menu expanded Button correctly', () => {
    const component = renderer.create(<Button expanded content="Menu expanded button" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a menu expanded primary Button correctly', () => {
    const component = renderer.create(<Button expanded primary content="Menu expanded primary button" />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a split Button correctly', () => {
    const component = renderer.create(<Button split icon="Add" content="Default split button" menu={buttonMenu} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a disabled split Button correctly', () => {
    const component = renderer.create(<Button split disabled icon="Add" content="Disabled split button" menu={buttonMenu} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a primary split Button correctly', () => {
    const component = renderer.create(<Button split primary icon="Add" content="Primary split button" menu={buttonMenu} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a disabled primary split Button correctly', () => {
    const component = renderer.create(
      <Button split disabled primary icon="Add" content="Disabled primary split button" menu={buttonMenu} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a split Button with its first action disabled correctly', () => {
    const component = renderer.create(
      <Button split primaryActionDisabled icon="Add" content="First action disabled split button" menu={buttonMenu} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a primary split Button with its first action disabled correctly', () => {
    const component = renderer.create(
      <Button split primaryActionDisabled primary icon="Add" content="First action disabled primary split button" menu={buttonMenu} />
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a multiline Button correctly', () => {
    const component = renderer.create(
      <Button icon="Share" menu={buttonMenu}>
        <Stack padding="8px 0" as="span" horizontalAlign="start">
          <Text>I am a compound multiline button.</Text>
          <Text variant="small">I can have a caption.</Text>
        </Stack>
      </Button>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });

  it('renders a Button inside a CommandBar correctly', () => {
    const component = renderer.create(<CommandBar items={[{ key: '0', text: 'Button 1', iconProps: { iconName: 'Upload' } }]} />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot;
  });
});
