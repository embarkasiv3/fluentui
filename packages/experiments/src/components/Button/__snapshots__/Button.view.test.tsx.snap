// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button view renders a default Button with content correctly 1`] = `
<button
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        background-color: #f4f4f4;
        border-color: transparent;
        border-radius: 0px;
        border-style: solid;
        border-width: 0px;
        box-sizing: border-box;
        color: #333333;
        cursor: default;
        display: inline-block;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        justify-content: center;
        margin-bottom: 0px;
        margin-left: 0px;
        margin-right: 0px;
        margin-top: 0px;
        min-height: 32px;
        min-width: 100px;
        outline: transparent;
        overflow: hidden;
        padding-bottom: 0px;
        padding-left: 0px;
        padding-right: 0px;
        padding-top: 0px;
        position: relative;
        text-align: center;
        text-decoration: none;
        user-select: none;
        vertical-align: baseline;
      }
      &::-moz-focus-inner {
        border: 0;
      }
      .ms-Fabric--isFocusVisible &:focus:after {
        border: 1px solid #ffffff;
        bottom: 1px;
        content: "";
        left: 1px;
        outline: 1px solid #666666;
        position: absolute;
        right: 1px;
        top: 1px;
        z-index: 1;
      }
      &:hover {
        background-color: #eaeaea;
        border-color: transparent;
        color: #212121;
      }
      &:hover:active {
        background-color: #eaeaea;
        border-color: transparent;
        color: #212121;
      }
      &:hover .ms-Icon-47 {
        color: #212121;
      }
      &:hover:active .ms-Icon-47 {
        color: #212121;
      }
  content="Default button"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  type="button"
>
  <span
    className=
        ms-Stack
        {
          align-items: center;
          box-sizing: border-box;
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          height: 100%;
          justify-content: center;
          padding-bottom: 8px;
          padding-left: 16px;
          padding-right: 16px;
          padding-top: 8px;
          width: auto;
        }
        & > * {
          text-overflow: ellipsis;
        }
        & > *:not(:first-child) {
          margin-left: 8px;
        }
        & > *:not(.ms-StackItem) {
          flex-shrink: 1;
        }
  >
    <span
      className=

          {
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            color: inherit;
            display: block;
            font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
            font-size: 14px;
            font-weight: 400;
            moz-osx-font-smoothing: grayscale;
            overflow: visible;
            text-overflow: ellipsis;
            webkit-font-smoothing: antialiased;
            white-space: nowrap;
          }
    >
      Default button
    </span>
  </span>
</button>
`;

exports[`Button view renders a disabled primary Button with content correctly 1`] = `
<button
  aria-disabled={true}
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        background-color: #f4f4f4;
        border-color: transparent;
        border-radius: 0px;
        border-style: solid;
        border-width: 0px;
        box-sizing: border-box;
        color: #a6a6a6;
        cursor: default;
        display: inline-block;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        justify-content: center;
        margin-bottom: 0px;
        margin-left: 0px;
        margin-right: 0px;
        margin-top: 0px;
        min-height: 32px;
        min-width: 100px;
        outline: transparent;
        overflow: hidden;
        padding-bottom: 0px;
        padding-left: 0px;
        padding-right: 0px;
        padding-top: 0px;
        position: relative;
        text-align: center;
        text-decoration: none;
        user-select: none;
        vertical-align: baseline;
      }
      &::-moz-focus-inner {
        border: 0;
      }
      .ms-Fabric--isFocusVisible &:focus:after {
        border: 1px solid #ffffff;
        bottom: 1px;
        content: "";
        left: 1px;
        outline: 1px solid #666666;
        position: absolute;
        right: 1px;
        top: 1px;
        z-index: 1;
      }
      &:hover {
        background-color: #f4f4f4;
        border-color: transparent;
        color: #a6a6a6;
      }
      &:hover:active {
        background-color: #f4f4f4;
        border-color: transparent;
        color: #a6a6a6;
      }
      &:hover .ms-Icon-57 {
        color: #a6a6a6;
      }
      &:hover:active .ms-Icon-57 {
        color: #a6a6a6;
      }
  content="Disabled primary button"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  type="button"
>
  <span
    className=
        ms-Stack
        {
          align-items: center;
          box-sizing: border-box;
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          height: 100%;
          justify-content: center;
          padding-bottom: 8px;
          padding-left: 16px;
          padding-right: 16px;
          padding-top: 8px;
          width: auto;
        }
        & > * {
          text-overflow: ellipsis;
        }
        & > *:not(:first-child) {
          margin-left: 8px;
        }
        & > *:not(.ms-StackItem) {
          flex-shrink: 1;
        }
  >
    <span
      className=

          {
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            color: inherit;
            display: block;
            font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
            font-size: 14px;
            font-weight: 400;
            moz-osx-font-smoothing: grayscale;
            overflow: visible;
            text-overflow: ellipsis;
            webkit-font-smoothing: antialiased;
            white-space: nowrap;
          }
    >
      Disabled primary button
    </span>
  </span>
</button>
`;

exports[`Button view renders a primary Button with content correctly 1`] = `
<button
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        background-color: #0078d4;
        border-color: transparent;
        border-radius: 0px;
        border-style: solid;
        border-width: 0px;
        box-sizing: border-box;
        color: #ffffff;
        cursor: default;
        display: inline-block;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        justify-content: center;
        margin-bottom: 0px;
        margin-left: 0px;
        margin-right: 0px;
        margin-top: 0px;
        min-height: 32px;
        min-width: 100px;
        outline: transparent;
        overflow: hidden;
        padding-bottom: 0px;
        padding-left: 0px;
        padding-right: 0px;
        padding-top: 0px;
        position: relative;
        text-align: center;
        text-decoration: none;
        user-select: none;
        vertical-align: baseline;
      }
      &::-moz-focus-inner {
        border: 0;
      }
      .ms-Fabric--isFocusVisible &:focus:after {
        border: 1px solid #ffffff;
        bottom: 1px;
        content: "";
        left: 1px;
        outline: 1px solid #666666;
        position: absolute;
        right: 1px;
        top: 1px;
        z-index: 1;
      }
      &:hover {
        background-color: #106ebe;
        border-color: transparent;
        color: #ffffff;
      }
      &:hover:active {
        background-color: #005a9e;
        border-color: transparent;
        color: #ffffff;
      }
      &:hover .ms-Icon-55 {
        color: #ffffff;
      }
      &:hover:active .ms-Icon-55 {
        color: #ffffff;
      }
  content="Primary button"
  onClick={[Function]}
  onKeyDown={[Function]}
  role="button"
  type="button"
>
  <span
    className=
        ms-Stack
        {
          align-items: center;
          box-sizing: border-box;
          display: flex;
          flex-direction: row;
          flex-wrap: nowrap;
          height: 100%;
          justify-content: center;
          padding-bottom: 8px;
          padding-left: 16px;
          padding-right: 16px;
          padding-top: 8px;
          width: auto;
        }
        & > * {
          text-overflow: ellipsis;
        }
        & > *:not(:first-child) {
          margin-left: 8px;
        }
        & > *:not(.ms-StackItem) {
          flex-shrink: 1;
        }
  >
    <span
      className=

          {
            -moz-osx-font-smoothing: grayscale;
            -webkit-font-smoothing: antialiased;
            color: inherit;
            display: block;
            font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
            font-size: 14px;
            font-weight: 400;
            moz-osx-font-smoothing: grayscale;
            overflow: visible;
            text-overflow: ellipsis;
            webkit-font-smoothing: antialiased;
            white-space: nowrap;
          }
    >
      Primary button
    </span>
  </span>
</button>
`;
